Add good key testing for ECDSA.

hlandau opened the pull request:
Fixes #158.

The need to implement curve checking is avoided by only supporting the curves that Go supports, which are the most popular curves. This is at least a big improvement over the current situation of not supporting any curves, and is a conservative approach security-wise.

Since crypto/elliptic supports only prime fields, the code assumes prime fields with regard to the public key validation advice found in NIST SP 800-56A.

osirisinferi commented:
In addition to your remark regarding the choice of curve checking: the [CA/Browser Forum Baseline Requirements](https://cabforum.org/wp-content/uploads/Baseline_Requirements_V1_3_1.pdf) also specify only the use of "NIST	P‐256,	P‐384,	or	P‐521" (section 6.1.5). So I'm guessing an extra argument for the hardcoded list of those three curves. (Although personally I would like to see a few `safecurves` as an option, but hey, can't have it al :stuck_out_tongue: No client support anyway..)

hlandau commented:
Yes, I'm very excited for Ed25519 support in TLS/X509/etc. Sadly the CFRG is still working on that, and even then it'll be a long wait until it's supported enough to be usable, I guess.

jsha commented:
Excellent, thanks for the pull request! I'm on vacation this week and next, but wanted to let you know I saw it and one of us will take a look.

jsha commented on core/good_key.go:
```diff
@@ -63,9 +64,80 @@ func GoodKey(key crypto.PublicKey) error {
+	//   Verify that key is not the point at infinity O.
+	if isAdditiveIdentity(key.Curve, key.X, key.Y) {
+		err = MalformedRequestError("Key x, y must not be the point at infinity")
+		log.Debug(err.Error())
```
The existing `log.Debug(err.Error())` in this function was a mistake - there's not enough context in the logged message, and errors returned from this function will get logged by the caller anyhow. Please remove these debug log statements. That should also allow you to move to a clearer `return MalformedRequestError(...)`, and remove the `log := blog.GetAuditLogger()` above.

jsha commented on core/good_key.go:
```diff
@@ -63,9 +64,80 @@ func GoodKey(key crypto.PublicKey) error {
+	log := blog.GetAuditLogger()
+
+	// Simply use a whitelist for now.
+	switch c.Params().Name {
```
The paranoid in me dislikes treating the Name field as authoritative here.  Since `CurveParams` is a struct, you can compare directly. This might look something like:

```
case "P-256" && params == elliptic.P256().Params():
    return nil
case "P-384" && params == elliptic.P384().Params():
   return nil
...
```

jsha replied:
Also, feedback from our technical advisory board is to not yet sign P521 keys, since support isn't there yet in the browsers.

osirisinferi replied:
Isn't there a risk of some chicken & egg scenario?

bifurcation replied:
As long as it's a quick copy/paste away, it doesn't seem like a huge problem to hold off.

jsha commented on core/good_key.go:
```diff
@@ -100,19 +172,29 @@ func GoodKeyRSA(key rsa.PublicKey) (err error) {
+	return nil
+}
+
+func checkSmallPrimes(i *big.Int) bool {
```
Document this function as to when it returns true or false. Or make it return an error, which makes it clear.

jsha commented on core/good_key.go:
```diff
@@ -63,9 +64,80 @@ func GoodKey(key crypto.PublicKey) error {
+		return err
+	}
+
+	// Assuming a prime field since a) we are only allowing such curves
```
Can you elaborate in this comment what impact it has on this particular line? In the unlikely event we start supporting non-prime curves in the future, what needs to change? We should probably document that next to the list of acceptable named curves, just to be safe.

jsha commented on core/good_key.go:
```diff
@@ -63,9 +64,80 @@ func GoodKey(key crypto.PublicKey) error {
+	}
+
+	// Ensure that public key is on the correct elliptic curve:
+	// verify that y**2 === x**3 - 3*x + B (mod P).
```
Shouldn't this be `Ax` instead of `3*x`?

bifurcation replied:
IIRC, In the curves we're talking about, it's always true that `A == 3`.

jsha commented:
Overall: great change, thanks for making it! I'd also appreciate if you could link in the comments to supporting documentation for this set of checks.

hlandau commented:
1. Disabled P521 for now. You can uncomment it safely if you wish.
2. Refactored error handling as requested.
3. Added comments to key validation.
4. Added comment to checkSmallPrimes. Leaving this function in was a mistake; I needed to use it when I was trying to implement curve validation, before I simply decided to use a whitelist, thus I refactored it into a function. But the code is certainly cleaner for it, so I'll leave it in. Commented.
5. WARNING: Merging this code is sufficient to cause boulder to issue certificates for ECDSA keys, assuming any HSM is willing to do so. Enabling ECDSA account keys will require a further change.

autoit4you commented on core/good_key.go:
```diff
@@ -54,37 +53,135 @@ func GoodKey(key crypto.PublicKey) error {
+		return err
+	}
+
+	// Key validation routine adapted from NIST SP800-56A § 5.6.2.5.
```
Note that in the current revision(Revision 2) of NIST SP800-56A from 2013 this moved to 5.6.2.3.2.
For reference: http://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-56Ar2.pdf

bifurcation commented:
@hlandau Could you please `git merge master` to update this branch?

hlandau commented:
Done.

hlandau commented:
Looks like travis ran into that DNS timeout issue, can someone restart it?

bifurcation commented on core/good_key.go:
```diff
@@ -54,37 +53,135 @@ func GoodKey(key crypto.PublicKey) error {
+	// Simply use a whitelist for now.
+	params := c.Params()
+	switch {
+	case params.Name == "P-256" && params == elliptic.P256().Params():
```
Isn't it duplicative to check the name separately from the rest of the struct?  Really, it seems like you could just inline this function in `GoodKeyECDSA`

hlandau replied:
I suppose. jsha seemed to want it paranoid. And I see no reason to inline it; it's clearer as a separate function, and operates on a different entity (elliptic.Curve).

bifurcation replied:
Cf. http://play.golang.org/p/GsnNIxqYr-

jsha replied:
Agreed that it doesn't make sense to inline.

@bifurcation: Good point that comparing the whole struct inherently also compares the name field. When I originally proposed this, I missed that the name field is part of the struct. I think your proposed change (switch over the whole params struct) is a nice one and improves clarity.

bifurcation commented on core/good_key.go:
```diff
@@ -54,37 +53,135 @@ func GoodKey(key crypto.PublicKey) error {
+		return err
+	}
+
+	// Key validation routine adapted from NIST SP800-56A § 5.6.2.3.2.
```
URL would be helpful here.

autoit4you replied:
http://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-56Ar2.pdf

bifurcation commented on core/good_key.go:
```diff
@@ -54,37 +53,135 @@ func GoodKey(key crypto.PublicKey) error {
+	// 5.6.2.3.2 and adapt this code appropriately.
+	params := key.Params()
+
+	//   Partial check of the public key for an invalid range in the EC group:
```
Remove leading spaces.

bifurcation commented on core/good_key.go:
```diff
@@ -54,37 +53,135 @@ func GoodKey(key crypto.PublicKey) error {
+	// The point at infinity (and additive identity) for all supported curves is
+	// (0,0). But we can implement this more generally by satisfying ourselves
+	// that the values provided are not the additive identity.
+	ax, ay := c.Add(x, y, x, y)
```
`2*P = 0` is not sufficient evidence that the point is the identity -- it could be a point of order two, which would be bad!  Rather than try to be fancy, I would just check for (0,0) inline above (with a caveat about the dependence on the curves), and delete this function.

bifurcation replied:
It would also be nice if `elliptic.Curve` had a method that would check for the identity on that curve.  Maybe consider submitting a patch upstream.

bifurcation commented on core/good_key_test.go:
```diff
@@ -84,3 +79,112 @@ func TestGoodKey(t *testing.T) {
+		test.AssertError(t, GoodKey(&public), "Should not have accepted key with point at infinity.")
+		test.AssertError(t, GoodKey(public), "Should not have accepted key with point at infinity.")
+	}
+}
```
Not seeing a test here for a point on the curve without the proper order.  Is it feasible to come up with an example?

SkateScout replied:
1) TestECDSANotOnCurveY so there is an check for an point not on the curve.
2) The check for additiveIdentity is ok, because 2 * (0,0) = (0,0) and if there would
   be any other point that is going to itself it is also not useable as an private key.

> > 2*P = 0 is not sufficient evidence that the point is the identity

> func isAdditiveIdentity(c elliptic.Curve, x, y *big.Int) bool {
> 	// The point at infinity (and additive identity) for all supported curves is
> 	// (0,0). But we can implement this more generally by satisfying ourselves
> 	// that the values provided are not the additive identity.
> 	ax, ay := c.Add(x, y, x, y)
> 	return ax.Cmp(x) == 0 && ay.Cmp(y) == 0
> }

bifurcation commented:
Overall, this looks like it's on the right track, modulo a couple of clean-up comments.  You could possibly dial back the verbosity of the comments one notch, e.g., just dropping all mention of non-prime-field curves.

Out of curiosity: What else is needed in order to enable ECDSA account keys?

hlandau commented:
The patch to enable ECDSA account keys is tiny. Still, it should be done separately. The main issue is probably getting test coverage.

jmhodges commented:
Ah, I haven't dug into this since I'm not one of the resident crypto folks. However, enabling ECDSA attempts as soon as this is deployed isn't going to work out well. We'll have folks that will try, get (at best) a cryptic error and, at worst, a panic from a mis-placed type conversion in the CA, (and also maybe the WFE?) where these keys are operated on. 

Please that a configurable setting that we can turn on later. This may require converting the GoodKey function into a struct with a method.

jmhodges commented:
(To be clear: we've already had folks attempt and fail to get an ECDSA cert, so that's not some hypothetical case. Clarity is super important for this because we don't know exactly how long it will take before we get a ECDSA signing key but its on the order of months and that's too long to have cryptic errors and dupe tickets made)

osirisinferi commented:
[quote](...) because we don't know exactly how long it will take before we get a ECDSA signing key (...)[/quote]
ECDSA certificates can be signed with a RSA key. No need to wait for new ECDSA keypairs on LE's end. I don't see the problem?

rmhrisk commented:
Yes, and this is quite common. I did a quick review and don't believe this would cause any issues.

T917820981726 commented:
+1
I'm one of the users trying to submit an ECDSA key CSR with Letsencrypt and got the "Invalid key in certificate request :: ECDSA keys not yet supported" error.
It would be really nice to be able to get ECDSA certificates as soon as possible, even if signed by RSA certificates (as other CAs already do). Thanks!

hlandau commented:
@jmhodges I grepped the whole source code for references to RSA and am satisfied there are no errant type assertions. I can use the code locally to issue ECDSA certificates. It all works. An RSA intermediate can sign an ECDSA cert.

hlandau commented:
Changed to check explicitly for point at infinity.

SkateScout commented:
@hlandau i am not sure if it is an good idea to check only for the nistp infinity point.
1) The AdditiveIdentity include the point for for nistp curves, so it is not worse.
     But for each curve an point that fulfill 2*P=P is unusable as key.
2) The error text should be either point of infinity or point of identity.
    "Public key does not have correct order" is missleading i think.

SkateScout commented:
3) Line 138  "n*Q = O iff n*Q is the point at infinity (see step 1)." should be "if" instead of "iff" i think.
4) Is there any good reason that "P-521" is currently excluded ?

osirisinferi commented:
> 4) Is there any good reason that "P-521" is currently excluded ?

Above @jsha stated the reason is lack of support in current browsers.

In my opinion not a very *good* reason, as ECDSA certificates currently can only be issued by manual CSR's, not somehow "selected" in the client itself. If the argument is "it would be confusing for users to generate a certificate which isn't supported by current browsers", then the choice (yes/no to issue a P-521 certificate) should be made somewhere else in the client if there's going to be a global option for ECDSA certificate issuing (i.e.: without a custom CSR). But until then IMHO there isn't really a good reason for disabeling P-521 in GoodKeyECDSA() and/or manual CSR's.

SkateScout commented:
1) Firefox at least in Alpha Support secp521r1
2) Chrome Desktop not
3) IE not
4) Java support it

rmhrisk commented:
Personally I would support secp521r1, other CAs do, but I can understant @jsha's possition.

osirisinferi commented:
> Personally I would support secp521r1, other CAs do, but I can understant @jsha's possition.

Personally, I don't understand why "companies" (not just LE) have such a "We know it better than those stupid, IQ-less (l)users, lets protect them for their own failures even before it even happened" attitude.. (Like a well known company with some fruity logo..) In my opinion, users should be offered a choice. At most they should be properly *warned* for some mishap waiting to happen (like breaking something important), but should have the option nonetheless.

rmhrisk commented:
@osirisinferi because they have to support the phone calls / emails and deal with associated PR.

SkateScout commented:
@jsha is it possible that you check the issue of signing P521 keys again with the tech board ?
Because firefox is supporting the P521 and as always said the certificate is not only for browsers.
But also used for other webapplication and even other protocols like smtp/imap/pop/irc so only to
look at browsers is an little short view. Also TOR.Browser which is based on older firefox versionsupport this curve. And with tor many people are more or less paranoid about security and like really high security grades.

osirisinferi commented:
@rmhrisk Well, have you looked at the [Community](community.letsencrypt.org) recently? I have a feeling that the guys who can generate their own ECDSA CSR know what they are doing, compared with all the braind**d "server admins" you sometimes encounter over there. Server admins who shouldn't be adminning a server :wink:

hlandau commented:
I don't see why P521 shouldn't be supported just because support is poor. I think it's @jsha's call, though.

rmhrisk commented:
Again, if I were asked I would advise to support. That said, your argument suggests none of the people with bad TLS configurations should be allowed to be admins :)

SkateScout commented:
Hi, i log the ssl hello packet of my clients to see what is supported.
If i check only for the support of the 3 curves that are discussed in these topic i get the following numbers.
grep -c secp521r1 sslHello.log > 28301
grep -c secp384r1 sslHello.log > 34278
grep -c secp256r1 sslHello.log > 34340
I know that this will change depending on the side. But i also know that i am be able to serve the certificate depended on what the client is able to talk.

jsha commented:
> WARNING: Merging this code is sufficient to cause boulder to issue certificates for ECDSA keys, assuming any HSM is willing to do so. Enabling ECDSA account keys will require a further change.

I agree with the comments above that we can sign ECDSA keys with our current RSA intermediate, and that's the plan. I believe the HSM shouldn't care what we are signing, since it is not aware of the content of the certs. Is there any reason to believe we would run into trouble there?

Also, even though we can issue certificates today, I'd like to have ECDSA support handled by a new config flag, so that we can enable support independently of deploy, in case there are any unexpected issues. I'm happy to implement that, or if you have time you can work it into this change.

Other than that, the only thing I think is needed here is @bifurcation's proposed switch statement change. Really excellent PR, thanks for the thorough documentation.

hlandau commented:
Removed explicit name check.
I don't see any reason why having an RSA intermediate sign ECDSA subjects should be an issue.

rmhrisk commented:
HSM will not care.

hlandau commented:
The issue with making this togglable is that GoodKey is stateless. So either you make it take some sort of context (bleh...) or you configure it via a global, which would be out of kilter with boulder configuration norms as far as I can tell.

Personally I'd use a global here, but that's just my personal preference, as in what I'd do in my own project. What do you think?

jmhodges commented:
Ah, please make a struct! We prefer tests to be easy to understand as locally as possible. Requiring one-off globals to be flipped on and off at a distance leads to mistakes and, often enough, races in tests. Totally understand that the constructors right now are rough and their type signatures are long, but we're planning on fixing that up.

Tangentially, just because the HSM should like it doesn't mean something (bugs in the code, a buggy bit of hardware, some inaccurate assumptions about the user base, etc.) along the way won't go wrong. It's a core bit of tech and worthy of a configuration flag. 

We totally think we could make our config stuff like this easier to manage, but we've not had the time to implement a real feature flagging system in boulder (like Decider, [flipper](https://github.com/jnunemaker/flipper), and a bunch of other systems). It's desirable, but not done.

**If you're already convinced that the config flags are a good idea, ignore this next paragraph.** It's mostly for the folks playing along at home and wondering why we're such sticklers about it.

We've got a whole section on [deployability in our developer guildelines](https://github.com/letsencrypt/boulder/wiki/Boulder-Development#deployability). We do this kind of config flag work a lot. New RPCs get them, what challenges are allowed are config flagged, even turning on Google SafeBrowsing checks got flags, and so on. Having config flags for this allows us to separate release of a code path from its deployment. Separating release from deploy means deploys get less risky for a few different reasons. We gain more fine-grained control over when and what kinds of risk we want to take on (instead of all at once with whatever other changes are deployed). That also means we don't  have to roll back all the way the past that build version (or making revert patch under pressure) if the code path turns out to be a problem. In the same way, deploying and testing staging gets easier because we don't have to be on Exactly The Right Build Version in order to know something was turned on and only it was turned on when we want folks to test it in a prod-like setting.

hlandau commented:
Key policy configuration added. Test configuration file updated.

hlandau commented:
This is out of scope for this PR. I accidentally committed some local changes I use for testing. But yes, ip6.arpa should be added, actually arpa as a whole should be added.

SkateScout commented:
Hi, i think the discussion about blacklisting arpa is not related to EC-Key integration.

jmhodges commented:
Ah, I think I wasn't clear enough! I meant a struct with a method on it that would take over the job of GoodKey.

It might be a wise idea for it to go into policy, too.

jmhodges commented on core/good_key.go:
```diff
@@ -38,53 +38,163 @@ var (
 
+// KeyPolicy etermines which types of key may be used with various boulder
+// operations.
+type KeyPolicy struct {
```
Please put this config object in cmd, instead. We're trying hard to keep our config objects out of the production code paths.

Also, it's name is a bit vague. We have lots of keys and policy things. How about "AllowedSigningAlgos"?

Then you can also drop the "allow" from all of these things names.

hlandau commented:
How can KeyPolicy go in cmd when GoodKey is in core?

jmhodges commented:
You don't use KeyPolicy elsewhere. Make another struct with a method that takes over for GoodKey and just use KeyPolicy (renamed) to fill in the slots of that new struct.

We're trying to keep JSON config objects out of the actual code because they have a bad tendency to stop fitting the model you need in the code but folks will try to work around them since they have gotten everywhere.

jmhodges commented:
We can put the GoodKey replacement struct in core for now. That might make this easier.

We're trying to strip core down because it's pretty overloaded. Way too much stuff in it.

hlandau commented:
@jmhodges How's this?

Probably should squash all these commits at some point.

jsha commented on core/good_key.go:
```diff
@@ -38,53 +38,163 @@ var (
+	case policy.AllowECDSANISTP521 && params == elliptic.P521().Params():
+		return nil
+	default:
+		return MalformedRequestError("ECDSA curve must be P256 or P384, not " + params.Name)
```
This should now be updated to `fmt.Sprintf("ECDSA curve %s not allowed.", params.Name)` so it makes sense regardless of what's currently enabled.

jsha commented on ca/certificate-authority.go:
```diff
@@ -65,6 +65,7 @@ type CertificateAuthorityImpl struct {
 	SA             core.StorageAuthority
 	PA             core.PolicyAuthority
 	Publisher      core.Publisher
+	KeyPolicy      core.KeyPolicy
```
Could you make this field unexported and set it using an argument in the CA constructor? As you can see, we set some of the fields in CA and RA by making them exported and assigning to them after construction, but we've found that that's an anti-pattern and have been meaning to fix it: https://github.com/letsencrypt/boulder/wiki/Config-plan.

hlandau commented:
@jsha Done.

jsha commented:
LGTM, thanks!

jmhodges commented:
LGTM

jmhodges commented:
FYI, looks like there's more work to do for ECDSA support to go in: https://github.com/letsencrypt/boulder/issues/1384

The boulder team is not likely to be able to get to this soon but we'll gladly review PRs for it. We've got other priorities right now, unfortunately.

Darkudorus commented:
Hi, I want to use P-521 for the latest Firefox (which also supports AES256-GCM), Safari and a mobile App. But Let's Encrypt currently supports only up to P-384 (which has equal security as RSA 7680 bit, which would have the same security level as AES192). But I want full security: AES256 == RSA 15360 bit == P-521. acme.sh is in use and supports P-521. Is there any way, that you can enable it at least on the testing servers for manual use? We are using DNSSEC + DANE + HSTS + HPKP. So P-521 would be a perfect completion. Thank you.
